DDP-7103 options groups for picklist study builder

aglushko-broad opened the pull request:
![image](https://user-images.githubusercontent.com/82094572/142433290-0e6df3b6-88e8-483f-a560-1c45b0e685a5.png)

Commit:
DDP-7103 add picklist question editor

Commit:
DDP-7103 add mock options

Commit:
DDP-7103 add label

Commit:
DDP-7103 picklist options editor

Commit:
DDP-7103 option groups

Commit:
DDP-7103 option groups

Commit:
Merge branch 'develop' into DDP-7103-study-builder-picklist

# Conflicts:
#	ddp-workspace/projects/study-builder-ui/src/app/app.module.ts
#	ddp-workspace/projects/study-builder-ui/src/app/components/manage-list/manage-list.component.html
#	ddp-workspace/projects/study-builder-ui/src/app/components/manage-list/manage-list.component.scss
#	ddp-workspace/projects/study-builder-ui/src/app/components/manage-list/manage-list.component.ts
#	ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-block/picklist-question-block.component.ts
#	ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.html
#	ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.ts

Commit:
DDP-7103 option groups

Commit:
DDP-7103 unique option stable ids and group ids validation

alextanlan commented on ddp-workspace/projects/study-builder-ui/src/app/components/picklist-group-editor/picklist-group-editor.component.ts:
```diff
@@ -0,0 +1,92 @@
+            this.groupForm.get('stableId').setErrors({stableIdIsNotUnique: true});
+        } else {
+            this.groupForm.get('stableId').setErrors(null);
+        }
```
_**#minor**_

Can be shortened like 
```
const error = stableIdIsNotUnique ? {stableIdIsNotUnique: true} : null;
this.groupForm.get('stableId').setErrors(error);
```

aglushko-broad replied:
fixed

alextanlan commented on ddp-workspace/projects/study-builder-ui/src/app/components/picklist-option-editor/picklist-option-editor.component.ts:
```diff
@@ -29,6 +29,14 @@ export class PicklistOptionEditorComponent implements OnDestroy {
+        } else {
+            this.optionForm.get('stableId').setErrors(null);
+        }
+    }
```
see [comment above](https://github.com/broadinstitute/ddp-angular/pull/953/files#r756146752)

aglushko-broad replied:
fixed

alextanlan commented on ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.ts:
```diff
@@ -83,11 +83,18 @@ export class PicklistQuestionEditorComponent implements OnInit, OnDestroy {
+    public allowCreateOptionsGroups(): boolean {
+        const selectMode = this.formGroup.get('selectMode').value;
+        const renderMode = this.formGroup.get('renderMode').value;
+        return selectMode === 'MULTIPLE' && renderMode === 'LIST' || renderMode === 'CHECKBOX_LIST' || renderMode === 'AUTOCOMPLETE';
```
_**#minor**_

can be shortened like
```
return selectMode === 'MULTIPLE' && ['LIST', 'CHECKBOX_LIST', 'AUTOCOMPLETE'].includes(renderMode);
```

aglushko-broad replied:
not really, the logical operator will work like 
"(selectMode === 'MULTIPLE' && renderMode === 'LIST') || renderMode === 'CHECKBOX_LIST' || renderMode === 'AUTOCOMPLETE';"

alextanlan replied:
My bad, agree.

MocanaAtBroad commented:
Looks good. Just a couple of little things I noticed.

MocanaAtBroad commented:
I created two picklist questions, added groups and options to each.
But when I clicked one question and then the other (in the preview panel), the editor panel was getting updated (the question prompt indicated the selected question text). But the panel with the options and groups was stuck showing the options and groups for one question/was not being updated with the selection.

MocanaAtBroad commented:
This I don't think is a bug, but did surprise me: When we create a picklist (checkbox list) with groups that have options, the groups are listed one on top of the other, but the options for each group are listed horizontally. I guess I was expecting that the default layout was to keep everything in vertical orientation. Could you just check that this indeed what we are doing by default (again, doubt this is a bug, but just something that surprised me)
![Screen Shot 2021-11-29 at 6 11 14 PM](https://user-images.githubusercontent.com/29984380/143957267-da388f3a-2256-491a-84ab-e1bc19555949.png)
.

aglushko-broad commented:
> This I don't think is a bug, but did surprise me: When we create a picklist (checkbox list) with groups that have options, the groups are listed one on top of the other, but the options for each group are listed horizontally. I guess I was expecting that the default layout was to keep everything in vertical orientation. Could you just check that this indeed what we are doing by default (again, doubt this is a bug, but just something that surprised me) ![Screen Shot 2021-11-29 at 6 11 14 PM](https://user-images.githubusercontent.com/29984380/143957267-da388f3a-2256-491a-84ab-e1bc19555949.png) .

![image](https://user-images.githubusercontent.com/82094572/144052243-77977165-789e-417c-aa30-d9de819423f1.png)
that's what I was able to find. In MPC project we put options into 2 columns. I'll copy all styles  for .ddp-picklist-group-container from MPC

Commit:
DDP-7103 fix comments

aglushko-broad commented:
> I created two picklist questions, added groups and options to each. But when I clicked one question and then the other (in the preview panel), the editor panel was getting updated (the question prompt indicated the selected question text). But the panel with the options and groups was stuck showing the options and groups for one question/was not being updated with the selection.

it was a bug, thanks, fixed!

Commit:
Merge branch 'develop' into DDP-7103-study-builder-picklist

# Conflicts:
#	ddp-workspace/projects/study-builder-ui/src/app/app.module.ts

Commit:
Merge branch 'develop' into DDP-7103-study-builder-picklist

# Conflicts:
#	ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.ts

MocanaAtBroad reviewed (approved):
Looks very good!
